src/metadata/trackmetadata.cpp

```diff
@@ -0,0 +1,80 @@
+#include "metadata/trackmetadata.h"
+
+#include "util/math.h"
+
+namespace Mixxx {
+
+/*static*/ const double TrackMetadata::BPM_UNDEFINED = 0.0;
+/*static*/ const double TrackMetadata::BPM_MIN = 0.0; // exclusive lower bound
+/*static*/ const double TrackMetadata::BPM_MAX = 300.0; // inclusive upper bound
+
+/*static*/ const double TrackMetadata::REPLAYGAIN_UNDEFINED = 0.0f;
+/*static*/ const double TrackMetadata::REPLAYGAIN_MIN = 0.0f; // exclusive lower bound
+/*static*/ const double TrackMetadata::REPLAYGAIN_0DB = 1.0f;
+
+TrackMetadata::TrackMetadata() :
+        m_channels(0),
+        m_sampleRate(0),
+        m_bitrate(0),
+        m_duration(0),
+        m_bpm(BPM_UNDEFINED),
+        m_replayGain(REPLAYGAIN_UNDEFINED) {
+}
+
+double TrackMetadata::parseBpmString(const QString& sBpm) {
+    if (sBpm.trimmed().isEmpty()) {
+        return BPM_UNDEFINED;
+    }
+    bool bpmValid = false;
+    double bpm = sBpm.toDouble(&bpmValid);
+    if ((!bpmValid) || (BPM_MIN > bpm)) {
+        qDebug() << "Failed to parse BPM:" << sBpm;
+        return BPM_UNDEFINED;
+    }
+    while (bpm > BPM_MAX) {
+        bpm /= 10.0;
+    }
+    return bpm;
+}
+
+bool TrackMetadata::setBpmString(const QString& sBpm) {
+    const double bpm = parseBpmString(sBpm);
+    if (BPM_UNDEFINED != bpm) {
+        setBpm(parseBpmString(sBpm));
+        return true;
+    } else {
```

Comment:
This is exactly how the existing code I found worked. As before I tried to keep the semantics even if I didn't agree ;) I will change the parseBpmString() function so that 0 is actually set when parsed correctly from the tag.